### ISSUE TITLE
Slow indexing in preprocessing - potential improvement

### ISSUE THREAD

**Issue by Peter9192** [WAM2layers/WAM2layers#138]
The preprocessing is slow and seems to leak memory when executed in parallel with dask. In analyzing what's going on, I found a big difference in performance between the following two snippets:

```py
boundary = 111
lower_layer = dp.level > boundary
upper_layer = ~lower_layer
s_lower = cw.where(lower_layer).sum(dim="level")
s_upper = cw.where(upper_layer).sum(dim="level")
```

```py
boundary = 111
idx = dp.level.searchsorted(boundary, side='right')
upper = np.s_[:, :idx, :, :]
lower = np.s_[:, idx:, :, :]
s_lower = cw[lower].sum(dim="level")
s_upper = cw[upper].sum(dim="level")
```

For my test case (so absolute timing is meaningless, but the difference is informative), the first took 13s, the second only 8s. That's quite substantial, I'd say.

However, changing this is only possible for the model level data, since the pressure-level data has a gridpoint-dependent boundary index.

**Comment by Peter9192** [comment]
Further investigation suggests why dask doesn't like this. This is the task graph for the second block:

![image](https://user-images.githubusercontent.com/17080502/198533931-83820356-a842-4f41-962b-ebe72560efee.png)

Notice how the graph consists of nicely independent blocks of tasks. 

This is the task graph for the first block:

![image](https://user-images.githubusercontent.com/17080502/198534179-7005926d-a015-4c61-922f-80582eb2da8a.png)

Here, all the tasks are connected!

**Comment by Peter9192** [comment]
This is the task graph for when the code is read from disk. This is after this suggested fix for the lower/upper layer indexing.

![image](https://user-images.githubusercontent.com/17080502/198546811-75498244-1f8e-4153-ba23-28c5e794250e.png)

So there's still some work to do in eliminating the dependencies across chunks.

**Comment by Imme1992** [comment]
#119

**Comment by Imme1992** [comment]
Implemented in #119 Global_runs pull request which is merged with the main
